docs/proposals/component-registration.md

```diff
@@ -0,0 +1,363 @@
+  - [Next Steps](#next-steps)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Abstract
+
+Kubernetes was designed to be extensible from the start, but one factor currently limits that capability: the hardcoding of what defines a Kubernetes cluster. To reach enterprise-class capabilities (high availability, notifications, self-healing, reporting, and maintainability) Kubernetes first needs to support dynamic component registration and readiness probes.
+
+
+## Use Cases
+
+This proposal primarily covers features that can be achieved internally by Kubernetes itself, but it's clear that there are also some desirable features that must be provided by external deployment, management, or monitoring systems. These external systems, however, require some features within Kubernetes to enable them. So while not all of these use cases can be solved by Kubernetes alone, we do want to make them possible.
+
+### Deployment
+
+As a k8s operator, I want a kubectl command that allows me to validate that the cluster has been deployed and is fully functional.
+
+As a k8s operator, I want a kubectl command that allows me to check which components are deployed/installed on a specific cluster in order to infer which features it supports.
+
+Note: Inferring features from components isn't very useful with vanilla k8s, but becomes more useful as 3rd party or optional components are added (e.g. OpenShift) and/or if components-controller is split into multiple components.
+
+### Debugging
+
+As a k8s operator, I want a single endpoint (and kubectl command) where I can see the current condition of the cluster so that I can debug observed failures (dropped connections, unresponsiveness, etc.).
+
+Note: The endpoint should contain cause where possible, but at least the current condition of the components.
+
+### Notification
+
+As a k8s operator, I want to be notified when the cluster requires human intervention so that I can intervene.
+
+As an external notification system (e.g. PagerDuty), I want an endpoint that responds to regular requests with the current cluster condition so that I can determine (with my own rule set) when to notify the k8s operator.
+
+Note: What condition requires human intervention is not a rule set that can be hard-coded, is likely to be different for each cloud platform, changes based on component configuration, and is likely to require a time-series database.
+
+### Reporting
+
+As a k8s operator, I want to see a report/graph of the readiness/uptime/availability of the cluster over time so that I can make/satisfy SLAs.
+
+Note: At least requires a time-series database and probably a reporting engine.
+
+### Self-Healing
+
+As an k8s operator, I want the cluster to self-heal when individual components crash so that I don't get notified or need to intervene.
+
+As a local process monitoring tool (e.g. monit), I want to be able to determine the condition of the component process I am monitoring to determine if it needs to be restarted.
```

Comment:
Agreed with @lavalamp, kubelet already does liveness/readiness probing. The "component controller" in your [PR](https://github.com/mesosphere/kubernetes/blob/b66eb7b7d1e1abe4a3434e7dbac8321c008f5992/pkg/controller/component/component_controller.go) is doing the same thing and IMO is less desirable.

Apologize if there is already a consensus below since I haven't got that far.